### ISSUE TITLE
Docker stuck on starting

### ISSUE THREAD

**Issue by Hexta** [docker/docker#31479]
**Description**
Docker stuck on starting. Log contains dozens of duplicate lines like:
```
msg="containerd: notify OOM events" error="open /proc/965558/cgroup: no such file or directory"
```
**Steps to reproduce the issue:**
1. Start containers
2. Restart docker few times

**Describe the results you received:**
Docker stuck on starting.

**Describe the results you expected:**
Docker started.

**Output of `docker version`:**

```
Client:
 Version:      1.13.1
 API version:  1.26
 Go version:   go1.7.5
 Git commit:   092cba3
 Built:        Wed Feb  8 06:50:14 2017
 OS/Arch:      linux/amd64

Server:
 Version:      1.13.1
 API version:  1.26 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   092cba3
 Built:        Wed Feb  8 06:50:14 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```
Containers: 196
 Running: 188
 Paused: 0
 Stopped: 8
Images: 27
Server Version: 1.13.1
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 524
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: aa8187dbd3b7ad67d8e5e3a15115d3eef43a7ed1
runc version: 9df8b306d01f59d3a8029be411de015b7304dd8f
init version: 949e6fa
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.9.9-040909-generic
Operating System: Ubuntu 16.04.2 LTS
OSType: linux
Architecture: x86_64
CPUs: 80
Total Memory: 62.82 GiB
Name: k8s-node-10
ID: MI25:MVTM:NEP2:JKN5:4JZC:EU7P:GCTL:VLW5:XVOT:GXG4:TANC:SLZ6
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Comment by cpuguy83** [comment]
Can you provide the full logs by chance?

ping @mlaventure

**Comment by thaJeztah** [comment]
I see you're not running the default kernel for Ubunt 16.04 (which runs on 4.4) not sure if that makes a difference here?

    Kernel Version: 4.9.9-040909-generic

**Comment by Hexta** [comment]
@cpuguy83 https://gist.github.com/Hexta/6685404364215eac4551f9267e3d2b56

**Comment by Hexta** [comment]
@thaJeztah I`m not 100% sure but I believe that this is not a specific 4.9.9 kernel problem.

**Comment by thaJeztah** [comment]
Yeah, I'm not sure it's related, just noticed it. Could you try running the check-config script to check if there's possibly something with the kernel configuration (just to be sure): https://raw.githubusercontent.com/docker/docker/master/contrib/check-config.sh ?

**Comment by Hexta** [comment]
https://gist.github.com/Hexta/f9c30b56e18958bc8e3a33b52d4df11d

**Comment by cpuguy83** [comment]
Is the system out of memory?

**Comment by Hexta** [comment]
@cpuguy83 No, "notify OOM events" confused me too.

**Comment by mlaventure** [comment]
@Hexta can you put the daemon in debug mode and restart it?

right now I'm thinking `dockerd` and/or `containerd` died with a lot of containers running which happened to have died before `containerd` restarted. 

The OOM message seems to come from the restore  path when `containerd` is trying to "reattach" to those container.

It looks like that period is taking too long causing docker to think `containerd` is dead (this is because the grpc connection is only setup after `containerd` is done restoring "known" containers)

**Comment by Hexta** [comment]
@mlaventure https://gist.github.com/Hexta/82cf6c14177ee732d03e3a4071003d7a

**Comment by mlaventure** [comment]
@Hexta it does look like my intuition is correct.

Do you think you'd be able to test this patch: https://gist.github.com/mlaventure/a84ac3e6a9dec240d6481e9060ebfdb4 and see if it fixes your issue?

**Comment by Hexta** [comment]
@mlaventure Thanks! It seems to work.
There are logs https://gist.github.com/Hexta/afe160e3126a35306543a73b6babe954

**Comment by mlaventure** [comment]
@Hexta  thanks for testing, I'll make a PR for it

**Comment by thaJeztah** [comment]
will be fixed in 17.03.1 by https://github.com/docker/docker/pull/31662
